### PR TITLE
Fixed #6707 -- Added RelatedManager.set() and made descriptors' __set__ use it.

### PR DESCRIPTION
**Pull request by loic**

### CODE REVIEW

File: docs/ref/models/relations.txt [hunk 1]
```diff
@@ -135,12 +135,28 @@ Related objects reference
...
         accepts the ``bulk`` keyword argument.
 
+    .. method:: set(obj1, [obj2, ...])
+
+    .. versionadded:: 1.8
+
+        Replace the set of related objects::
+
+            >>> new_list = [obj1, obj2, obj3]
+            >>> e.related_set.set(new_list)
```
Comment (collinanderson) [inline docs/ref/models/relations.txt]: This is the API I would expect, but why is it implemented as .set(*object_list) rather than .set(object_list)?

Comment (collinanderson) [inline docs/ref/models/relations.txt]: Ohh, probably to be consistent with .add(*object_list)? Still, it seems to me set(object_list) would be the most useful.

Comment (loic) [inline docs/ref/models/relations.txt]: For consistency yes, I'd be quite reluctant breaking it.

File: docs/ref/contrib/admin/index.txt [hunk 2]
```diff
@@ -2206,6 +2206,10 @@ model to represent the many-to-many relationship, you must tell Django's admin
 to *not* display this widget - otherwise you will end up with two widgets on
 your admin page for managing the relation.
 
+Note that when using this technique the :data:`m2m_changed` signals aren't
+triggered. This is because as far as the admin is concerned, ``through`` is
+just a model with two foreign key fields rather than a many-to-many relation.
```
Comment (akaariai) [inline docs/ref/contrib/admin/index.txt]: This documents existing behavior, right?

**Comment by akaariai** [comment]
Looks good to me.

File: docs/ref/contrib/admin/index.txt (hunk 2 above)
Comment (loic) [inline docs/ref/contrib/admin/index.txt]: Yes, I've put it in a different commit.

File: docs/ref/models/relations.txt [hunk 3]
```diff
@@ -135,12 +135,31 @@ Related objects reference
         :class:`~django.db.models.ForeignKey`\s where ``null=True`` and it also
         accepts the ``bulk`` keyword argument.
 
+    .. method:: set(obj1, [obj2, ...])
+
+    .. versionadded:: 1.9
```
Comment (timgraham) [inline docs/ref/models/relations.txt]: indent

File: docs/ref/models/relations.txt [hunk 4]
```diff
@@ -135,12 +135,31 @@ Related objects reference
         :class:`~django.db.models.ForeignKey`\s where ``null=True`` and it also
         accepts the ``bulk`` keyword argument.
 
+    .. method:: set(obj1, [obj2, ...])
```
Comment (timgraham) [inline docs/ref/models/relations.txt]: should `clear=False` be listed here?

File: docs/ref/models/relations.txt [hunk 5]
```diff
@@ -135,12 +135,31 @@ Related objects reference
...
+            >>> new_list = [obj1, obj2, obj3]
+            >>> e.related_set.set(*new_list)
+
+        This method accepts a ``clear`` argument to control how to perform the
+        operation. If ``False`` (the default), the elements missing from the
+        new set are removed using ``remove()`` and only the new ones are added.
+        If ``clear=True``, the ``clear()`` method is called instead and the
+        whole set is added at once.
+
+        Note that ``set()`` being a compound operation it is subjected to race
```
Comment (timgraham) [inline docs/ref/models/relations.txt]: Note that since set is a compound operation, it is subject to...

File: docs/ref/models/relations.txt [hunk 6]
```diff
@@ -158,6 +177,11 @@ new iterable of objects to it::
...
-will first call ``clear()`` to disassociate any existing objects in the related
-set before adding the contents of ``new_list``. Otherwise the objects in
-``new_list`` will be added to the existing related object set.
+will first disassociate any existing objects in the related set before adding
+the contents of ``new_list``. Otherwise the objects in ``new_list`` will be
+added to the existing related object set.
+
+.. versionchanged:1.9
+
+In earlier versions direct assignment used to perform ``clear()`` followed by
```
Comment (timgraham) [inline docs/ref/models/relations.txt]: indent contents
In earlier versions,

File: docs/ref/models/relations.txt [hunk 7]
```diff
@@ -158,6 +177,11 @@ new iterable of objects to it::
...
+``add()``, it now performs a ``set()`` with the keyword argument ``clear=False``.
```
Comment (timgraham) [inline docs/ref/models/relations.txt]: . It now
or: "; it now"

File: docs/releases/1.9.txt [hunk 8]
```diff
@@ -127,7 +127,9 @@ Management Commands
 Models
 ^^^^^^
 
-* ...
+* Added the :meth:`django.db.models.fields.related.RelatedManager.set()` method
```
Comment (timgraham) [inline docs/releases/1.9.txt]: `RelatedManager.set() <django.db.models.fields.related.RelatedManager.set>` rather then rendering the entire path?

File: docs/releases/1.9.txt [hunk 9]
```diff
@@ -127,7 +127,9 @@ Management Commands
...
+  to the related managers created by ``ForeignKey`` and ``GenericForeignKey``
```
Comment (timgraham) [inline docs/releases/1.9.txt]: FK, GFK, and M2M (add commas)

File: docs/releases/1.9.txt [hunk 10]
```diff
@@ -165,7 +167,24 @@ Backwards incompatible changes in 1.9
     deprecation timeline for a given feature, its removal may appear as a
     backwards incompatible change.
 
-...
+Related set direct assignment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:ref:`direct assignment <direct-assignment>`) used to perform a ``clear()``
```
Comment (timgraham) [inline docs/releases/1.9.txt]: Direct assignment

File: docs/releases/1.9.txt [hunk 11]
```diff
@@ -165,7 +167,24 @@ Backwards incompatible changes in 1.9
...
+to track individual changes in many-to-many relations.
+
+Direct assignment now relies on the the new
+:meth:`django.db.models.fields.related.RelatedManager.set()` method on
+related managers which by default only processes changes between the
+existing related set and the one that's newly assigned. The previous behavior
+can be restored by replacing direct assignment by a call to ``set()`` with
+the keyword argument ``clear=True``.
+
+``ModelForm`` and therefore ``ModelAdmin`` internally rely on direct assignment
```
Comment (timgraham) [inline docs/releases/1.9.txt]: , and therefore ModelAdmin,

File: docs/releases/1.9.txt [hunk 12]
```diff
@@ -165,7 +167,24 @@ Backwards incompatible changes in 1.9
...
+for many-to-many relations and as a consequence will now display the new
```
Comment (timgraham) [inline docs/releases/1.9.txt]: will now display -> now use

File: tests/many_to_many/tests.py [hunk 13]
```diff
@@ -332,6 +332,45 @@ def test_remove(self):
             ])
         self.assertQuerysetEqual(self.a3.publications.all(), [])
 
+    def test_set(self):
+        self.p2.article_set.set(self.a4, self.a3)
+        self.assertQuerysetEqual(self.p2.article_set.all(),
+            [
```
Comment (timgraham) [inline tests/many_to_many/tests.py]: I'd prefer if you kept the same indent style as the above tests (but I guess you've copied the style of the surrounding code)

File: docs/ref/models/relations.txt (hunk 4 above)
Comment (loic) [inline docs/ref/models/relations.txt]: I wondered about it, we didn't do it for `bulk=True/False` so I didn't do it here. Should I add it?

File: tests/many_to_one_null/tests.py [hunk 14]
```diff
@@ -74,6 +74,22 @@ def test_remove_from_wrong_set(self):
         self.assertRaises(Reporter.DoesNotExist, self.r.article_set.remove, self.a4)
         self.assertQuerysetEqual(self.r2.article_set.all(), ['<Article: Fourth>'])
 
+    def test_set(self):
+        # Use manager.set() to allocate ForeignKey. Null is legal, so existing
+        # members of set that are not in the assignment set are set null
```
Comment (timgraham) [inline tests/many_to_one_null/tests.py]: of the set
set to null
adding period

File: docs/ref/models/relations.txt (hunk 4 above)
Comment (timgraham) [inline docs/ref/models/relations.txt]: I would

Comment (loic) [inline docs/ref/models/relations.txt]: Work for me, I'll add another commit which fixes it for bulk as well.

Comment (loic) [inline docs/ref/models/relations.txt]: Ok now I remember why I didn't add it for `bulk`, it's not available on all relational managers.

File: tests/many_to_many/tests.py (hunk 13 above)
Comment (loic) [inline tests/many_to_many/tests.py]: I don't like these either since they aren't pep8 compliant (E128) but the whole file (and the similar test packages) use them, so I'd like to keep them consistent.

File: tests/many_to_one_null/tests.py (hunk 14 above)
Comment (loic) [inline tests/many_to_one_null/tests.py]: This was copied from another test, fixed in both locations.

**Comment by loic** [comment]
Thinking about it a bit more, maybe `set()` should accept an iterable rather than `*objs`, @collinanderson raised the issue in https://github.com/django/django/pull/2500#discussion_r23572348.

I used `*objs` for consistency with `add()` and `remove()` but those two make sense for a single object, less so for `set()`. What do you think @akaariai, @carljm, @timgraham?

**Comment by timgraham** [comment]
Did you see https://code.djangoproject.com/ticket/13240 which proposes that for related methods?

**Comment by loic** [comment]
@timgraham wasn't aware of it. Can't say I like the idea of flattening all iterables like done in that patch, although I would accept special-casing when there is only 1 object and that object is an iterable (we should keep in mind that it would probably break models with `__iter__`).

Regarding `set(objs)` vs `set(*objs)`, I'm 40/60 in favor of the latter :

- When I think of it as a shorthand for `clear()` + `add()`, then I really want consistency with `add()`.
- When I think of a set that I just want to assign, then I find `set(objs`) nicer.

File: django/contrib/contenttypes/fields.py [hunk 15]
```diff
@@ -530,6 +522,27 @@ def _clear(self, queryset, bulk):
...
+        def set(self, *objs, **kwargs):
+            clear = kwargs.pop('clear', False)
+
+            db = router.db_for_write(self.model, instance=self.instance)
+            with transaction.atomic(using=db, savepoint=False):
+                if clear:
+                    self.clear()
+                    self.add(*objs)
+                else:
+                    old_objs = set(self.using(db).all())
```
Comment (MarkusH) [inline django/contrib/contenttypes/fields.py]: Care to open a ticket and add a comment so that it won't be forgotten?

**Comment by carljm** [comment]
I also think `.set(objs)` is better than `.set(*objs)`. I think the strongest parallel for `.set()` isn't `.add()`, it's "what you retrieve", which is an iterable.

(It's a tangent, but I also don't like methods that accept either and magically flatten iterables.)

**Comment by loic** [comment]
> I think the strongest parallel for .set() isn't .add(), it's "what you retrieve", which is an iterable.

That's very true.

**Comment by charettes** [comment]
I'd also argue that you're more likely to use `add()` for a single object than you are for `set()`. Most of the time you'll pass a collection of items when using `set()` since it's the many part of the relationship.

I think say `set()` shares more with `bulk_create()` than `add()` in this regard.

File: docs/ref/models/relations.txt (hunk 1 above)
Comment (loic) [inline docs/ref/models/relations.txt]: For posterity: I've changed my mind on this and consensus emerged in favor of `set(objs)` instead of `set(*objs)`.
